File: pluk/src/adapters/linear/resolve.ts
import { linearGraphQL } from "./client.js";

interface TeamNode {
  id: string;
  key: string;
  name: string;
}

interface UserNode {
  id: string;
  name: string;
  email: string | null;
}

function cap(items: string[]): string {
  const shown = items.slice(0, 10).join(", ");
  return items.length > 10 ? `${shown}, …` : shown;
}

export async function resolveTeam(apiKey: string, keyOrName: string): Promise<TeamNode> {
  const data = await linearGraphQL<{ teams: { nodes: TeamNode[] } }>(apiKey, `{ teams { nodes { id key name } } }`);
  const want = keyOrName.trim().toLowerCase();
  const exact = data.teams.nodes.filter((t) => t.key.toLowerCase() === want || t.name.toLowerCase() === want);
  if (exact.length === 1) return exact[0]!;
  if (exact.length > 1) {
    const list = exact.map((t) => `${t.key} (${t.name})`).join(", ");
    throw new Error(`Team "${keyOrName}" matches more than one team: ${list}. Pass the exact team key.`);
  }
  const known = data.teams.nodes.map((t) => `${t.key} (${t.name})`);
  throw new Error(`No team named "${keyOrName}". Known teams: ${cap(known)}.`);
}

export async function resolveUser(apiKey: string, emailOrName: string): Promise<UserNode> {
  const term = emailOrName.trim();
  const byEmail = term.includes("@");
  const filter = byEmail ? { email: { containsIgnoreCase: term } } : { name: { containsIgnoreCase: term } };
  const data = await linearGraphQL<{ users: { nodes: UserNode[] } }>(
    apiKey,
    `query($filter:UserFilter){ users(first: 50, filter: $filter){ nodes { id name email } } }`,
    { filter },
  );
  const want = term.toLowerCase();
  const exact = data.users.nodes.filter((u) =>
    byEmail ? (u.email ?? "").toLowerCase() === want : u.name.toLowerCase() === want,
  );
  if (exact.length === 1) return exact[0]!;
  if (exact.length > 1) {
    const list = exact.map((u) => (u.email ? `${u.name} <${u.email}>` : u.name)).join(", ");
    throw new Error(`Assignee "${emailOrName}" matches more than one user: ${list}. Pass a unique email or name.`);
  }
  const near = data.users.nodes.map((u) => (u.email ? `${u.name} <${u.email}>` : u.name));
  if (near.length) throw new Error(`No user matches "${emailOrName}". Near matches: ${cap(near)}.`);
  throw new Error(`No user matches "${emailOrName}".`);
}

export async function resolveState(apiKey: string, teamKey: string, name: string): Promise<{ id: string; name: string }> {
  const data = await linearGraphQL<{ workflowStates: { nodes: { id: string; name: string }[] } }>(
    apiKey,
    `query($filter:WorkflowStateFilter){ workflowStates(filter: $filter){ nodes { id name } } }`,
    { filter: { team: { key: { eq: teamKey } } } },
  );
  const want = name.trim().toLowerCase();
  const exact = data.workflowStates.nodes.filter((s) => s.name.toLowerCase() === want);
  if (exact.length === 1) return exact[0]!;
  if (exact.length > 1) {
    const list = exact.map((s) => s.name).join(", ");
    throw new Error(`State "${name}" matches more than one workflow state: ${list}. Pass the exact state name.`);
  }
  const names = data.workflowStates.nodes.map((s) => s.name);
  throw new Error(`No workflow state named "${name}" in team ${teamKey}. States: ${cap(names)}.`);
}

export async function resolveLabels(apiKey: string, names: string[]): Promise<string[]> {
  const data = await linearGraphQL<{ issueLabels: { nodes: { id: string; name: string }[] } }>(
    apiKey,
    `{ issueLabels(first: 250){ nodes { id name } } }`,
  );
  return names.map((raw) => {
    const want = raw.trim().toLowerCase();
    const exact = data.issueLabels.nodes.filter((l) => l.name.toLowerCase() === want);
    if (exact.length === 1) return exact[0]!.id;
    if (exact.length > 1) {
      const list = exact.map((l) => l.name).join(", ");
      throw new Error(`Label "${raw}" matches more than one label: ${list}. Pass the exact label name.`);
    }
    const known = data.issueLabels.nodes.map((l) => l.name);
    throw new Error(`No label named "${raw}". Existing labels: ${cap(known)}.`);
  });
}

File: crates/pluk-adapters/src/linear/resolve.rs
use serde_json::{json, Value};

use crate::error::AdapterError;
use super::client::linear_graphql;

fn cap(items: &[String]) -> String {
    let shown = items.iter().take(10).cloned().collect::<Vec<_>>().join(", ");
    if items.len() > 10 { format!("{shown}, …") } else { shown }
}

pub async fn resolve_team(api_key: &str, key_or_name: &str) -> Result<Value, AdapterError> {
    let data = linear_graphql(api_key, "{ teams { nodes { id key name } } }", json!({})).await?;
    let nodes = data.get("teams").and_then(|t| t.get("nodes")).and_then(|n| n.as_array()).cloned().unwrap_or_default();
    let want = key_or_name.trim().to_lowercase();
    let exact: Vec<&Value> = nodes.iter().filter(|t| {
        t.get("key").and_then(|k| k.as_str()).map(|k| k.to_lowercase() == want).unwrap_or(false)
            || t.get("name").and_then(|n| n.as_str()).map(|n| n.to_lowercase() == want).unwrap_or(false)
    }).collect();
    if exact.len() == 1 {
        return Ok(exact[0].clone());
    }
    if exact.len() > 1 {
        let list = exact.iter().map(|t| format!("{} ({})", t.get("key").and_then(|k| k.as_str()).unwrap_or("?"), t.get("name").and_then(|n| n.as_str()).unwrap_or("?"))).collect::<Vec<_>>().join(", ");
        return Err(AdapterError::new(format!("Team \"{key_or_name}\" matches more than one team: {list}. Pass the exact team key.")));
    }
    let known: Vec<String> = nodes.iter().map(|t| format!("{} ({})", t.get("key").and_then(|k| k.as_str()).unwrap_or("?"), t.get("name").and_then(|n| n.as_str()).unwrap_or("?"))).collect();
    Err(AdapterError::new(format!("No team named \"{key_or_name}\". Known teams: {}.", cap(&known))))
}

pub async fn resolve_user(api_key: &str, email_or_name: &str) -> Result<Value, AdapterError> {
    let term = email_or_name.trim().to_string();
    let by_email = term.contains('@');
    let filter = if by_email { json!({ "email": { "containsIgnoreCase": term } }) } else { json!({ "name": { "containsIgnoreCase": term } }) };
    let data = linear_graphql(api_key, "query($filter:UserFilter){ users(first: 50, filter: $filter){ nodes { id name email } } }", json!({ "filter": filter })).await?;
    let nodes = data.get("users").and_then(|u| u.get("nodes")).and_then(|n| n.as_array()).cloned().unwrap_or_default();
    let want = term.to_lowercase();
    let exact: Vec<&Value> = nodes.iter().filter(|u| {
        if by_email { u.get("email").and_then(|e| e.as_str()).map(|e| e.to_lowercase() == want).unwrap_or(false) } else { u.get("name").and_then(|n| n.as_str()).map(|n| n.to_lowercase() == want).unwrap_or(false) }
    }).collect();
    if exact.len() == 1 {
        return Ok(exact[0].clone());
    }
    if exact.len() > 1 {
        let list = exact.iter().map(|u| { let name = u.get("name").and_then(|n| n.as_str()).unwrap_or("?"); match u.get("email").and_then(|e| e.as_str()) { Some(e) => format!("{name} <{e}>"), None => name.to_string() } }).collect::<Vec<_>>().join(", ");
        return Err(AdapterError::new(format!("Assignee \"{email_or_name}\" matches more than one user: {list}. Pass a unique email or name.")));
    }
    let near: Vec<String> = nodes.iter().map(|u| { let name = u.get("name").and_then(|n| n.as_str()).unwrap_or("?"); match u.get("email").and_then(|e| e.as_str()) { Some(e) => format!("{name} <{e}>"), None => name.to_string() } }).collect();
    if !near.is_empty() {
        return Err(AdapterError::new(format!("No user matches \"{email_or_name}\". Near matches: {}.", cap(&near))));
    }
    Err(AdapterError::new(format!("No user matches \"{email_or_name}\".")))
}

pub async fn resolve_state(api_key: &str, team_key: &str, name: &str) -> Result<Value, AdapterError> {
    let data = linear_graphql(api_key, "query($filter:WorkflowStateFilter){ workflowStates(filter: $filter){ nodes { id name } } }", json!({ "filter": { "team": { "key": { "eq": team_key } } } })).await?;
    let nodes = data.get("workflowStates").and_then(|w| w.get("nodes")).and_then(|n| n.as_array()).cloned().unwrap_or_default();
    let want = name.trim().to_lowercase();
    let exact: Vec<&Value> = nodes.iter().filter(|s| s.get("name").and_then(|n| n.as_str()).map(|n| n.to_lowercase() == want).unwrap_or(false)).collect();
    if exact.len() == 1 { return Ok(exact[0].clone()); }
    if exact.len() > 1 {
        let list = exact.iter().filter_map(|s| s.get("name").and_then(|n| n.as_str())).collect::<Vec<_>>().join(", ");
        return Err(AdapterError::new(format!("State \"{name}\" matches more than one workflow state: {list}. Pass the exact state name.")));
    }
    let names: Vec<String> = nodes.iter().filter_map(|s| s.get("name").and_then(|n| n.as_str()).map(|s| s.to_string())).collect();
    Err(AdapterError::new(format!("No workflow state named \"{name}\" in team {team_key}. States: {}.", cap(&names))))
}

pub async fn resolve_labels(api_key: &str, names: &[String]) -> Result<Vec<String>, AdapterError> {
    let data = linear_graphql(api_key, "{ issueLabels(first: 250){ nodes { id name } } }", json!({})).await?;
    let nodes = data.get("issueLabels").and_then(|l| l.get("nodes")).and_then(|n| n.as_array()).cloned().unwrap_or_default();
    let mut ids = Vec::new();
    for raw in names {
        let want = raw.trim().to_lowercase();
        let exact: Vec<&Value> = nodes.iter().filter(|l| l.get("name").and_then(|n| n.as_str()).map(|n| n.to_lowercase() == want).unwrap_or(false)).collect();
        if exact.len() == 1 {
            ids.push(exact[0].get("id").and_then(|id| id.as_str()).unwrap_or("").to_string());
        } else if exact.len() > 1 {
            let list = exact.iter().filter_map(|l| l.get("name").and_then(|n| n.as_str())).collect::<Vec<_>>().join(", ");
            return Err(AdapterError::new(format!("Label \"{raw}\" matches more than one label: {list}. Pass the exact label name.")));
        } else {
            let known: Vec<String> = nodes.iter().filter_map(|l| l.get("name").and_then(|n| n.as_str()).map(|s| s.to_string())).collect();
            return Err(AdapterError::new(format!("No label named \"{raw}\". Existing labels: {}.", cap(&known))));
        }
    }
    Ok(ids)
}
